syntax = "proto3";

package common;
option go_package = "github.com/jalexanderII/zero_fintech/gen/Go/common";

import "common/common.proto";
import "google/protobuf/timestamp.proto";

message PaymentAction {
  // account to which payment shall be made
  string account_id = 1;
  // amount of payment
  double amount = 2;
  // Planned date of transaction
  google.protobuf.Timestamp transaction_date = 3;
  // status of payment action
  common.PaymentActionStatus status = 4;
}

message PaymentPlan {
  string payment_plan_id = 1;
  string user_id = 2;
  repeated string payment_task_id = 3;
  double amount = 4;
  double timeline = 5;
  common.PaymentFrequency payment_freq = 6;
  double amount_per_payment = 7;
  common.PlanType plan_type = 8;
  // Expected date the plan should be completed
  google.protobuf.Timestamp end_date = 9;
  // This Payment plan has not completed and is still active
  bool active = 10;
  // Current status of this plan
  common.PaymentStatus status = 11;
  // payment actions of the plan
  repeated PaymentAction payment_action = 12;
}

message PaymentPlanWName {
  string name = 1;
  string payment_plan_id = 2;
  string user_id = 3;
  repeated string payment_task_id = 4;
  double amount = 5;
  double timeline = 6;
  common.PaymentFrequency payment_freq = 7;
  double amount_per_payment = 8;
  common.PlanType plan_type = 9;
  // Expected date the plan should be completed
  google.protobuf.Timestamp end_date = 10;
  // This Payment plan has not completed and is still active
  bool active = 11;
  // Current status of this plan
  common.PaymentStatus status = 12;
  // payment actions of the plan
  repeated PaymentAction payment_action = 13;
}

// Response for CreatePaymentPlan
message PaymentPlanResponse {
  repeated PaymentPlanWName payment_plans = 1;
}

// CRUD Methods
message GetPaymentPlanRequest {
  string payment_plan_id = 1;
}

message ListPaymentPlanRequest {}

message ListUserPaymentPlansRequest {
  string user_id = 1;
}

message ListPaymentPlanResponse {
  repeated PaymentPlanWName payment_plans = 1;
}

message UpdatePaymentPlanRequest {
  string payment_plan_id = 1;
  PaymentPlanWName payment_plan = 2;
}

message DeletePaymentPlanRequest {
  string payment_plan_id = 1;
}

message DeletePaymentPlanResponse {
  common.DELETE_STATUS status = 1;
  PaymentPlanWName payment_plan = 2;
}